### ISSUE TITLE
FunctionReporter fails to parse complex function objects with functions as attributes

### ISSUE THREAD

**Issue by jayqi** [UptakeOpenSource/pkgnet#134]
For complex function objects that have other functions stored as attributes, `.parse_function` used by `FunctionReporter` will fail.

Here is a toy reproducible example, showing that the [`as.list` call](https://github.com/UptakeOpenSource/pkgnet/blob/a571b8e8a0cfa1a7c45cc83a08d95f46ce2e40bb/R/PackageFunctionReporter.R#L272) used in `.parse_function` would fail.
``` r
foo <- function() {"foo"}
bar <- function() {"bar"}
my_special_func_creator <- function (f) {
    structure(f
              , class = "my_special_func_type"
              , attribute_that_is_a_func = foo
    )
}
my_special_func <- my_special_func_creator(bar)
as.list(my_special_func)
#> Error in as.vector(x, "list"): cannot coerce type 'closure' to vector of type 'list'
```
This is because we are storing another function `foo` as an attribute of the function object that is bound to `my_special_func`, and `as.list` has trouble when it gets to it.

This is a design pattern that gets used by, for example, the [`decorator` package](https://github.com/klmr/decorator):
``` r
library(decorator)
twice = decorator %@% function (f) {
    function (...) {
        f(...)
        f(...)
    }
}
as.list(twice)
#> Error in as.vector(x, "list"): cannot coerce type 'closure' to vector of type 'list'
attributes(twice)
#> $srcref
#> function (f) {
#>     function (...) {
#>         f(...)
#>         f(...)
#>     }
#> }
#> 
#> $class
#> [1] "decorator" "decorated"
#> 
#> $decorators
#> $decorators$decorator
#> decorator
```
Decorators created by `decorator` have the function `decorator::decorator` stored as an attribute. 

<sup>Created on 2018-11-21 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by jayqi** [comment]
Addressed by #135

**Comment by jayqi** [comment]
Explanation is updated in the OP.  Previous incorrect explanation below:

> Here is a toy reproducible example, showing that the [`as.list` call](https://github.com/UptakeOpenSource/pkgnet/blob/a571b8e8a0cfa1a7c45cc83a08d95f46ce2e40bb/R/PackageFunctionReporter.R#L272) used in `.parse_function` would fail.
> ``` r
> foo <- function() {"foo"}
> bar <- function() {"bar"}
> my_special_func_creator <- function (f) {
>     structure(f
>               , class = "my_special_func_type"
>               , attribute_that_is_a_func = foo
>     )
> }
> my_special_func <- my_special_func_creator(bar)
> as.list(my_special_func)
> #> Error in as.vector(x, "list"): cannot coerce type 'closure' to vector of type 'list'
> ```
> This is because we are storing another function `foo` as an attribute of the function object that is bound to `my_special_func`, and `as.list` has trouble when it gets to it.
> 
> This is a design pattern that gets used by, for example, the [`decorator` package](https://github.com/klmr/decorator):
> ``` r
> library(decorator)
> twice = decorator %@% function (f) {
>     function (...) {
>         f(...)
>         f(...)
>     }
> }
> as.list(twice)
> #> Error in as.vector(x, "list"): cannot coerce type 'closure' to vector of type 'list'
> attributes(twice)
> #> $srcref
> #> function (f) {
> #>     function (...) {
> #>         f(...)
> #>         f(...)
> #>     }
> #> }
> #> 
> #> $class
> #> [1] "decorator" "decorated"
> #> 
> #> $decorators
> #> $decorators$decorator
> #> decorator
> ```
> Decorators created by `decorator` have the function `decorator::decorator` stored as an attribute. 
> 
> <sup>Created on 2018-11-21 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>
